### PR TITLE
extract manual optimization loop

### LINKED ISSUE #9283: RFC: Manual Optimization Loop
## Proposed refactoring or deprecation

After refactoring and extracting the automatic optimization into its own loop in #9191, what remains is the manual optimization logic in the `TrainingBatchLoop`. This RFC proposes to extract manual optimization into its own loop as well.

### Motivation

Loop customization has been made more feasible and accessible with the extraction of the optimizer loop in #9191. The disentangling of both modes is increasing code readability, maintainability and extensibility in the context of loop customization. The next logical step is to do the same for manual optimization, which would only encompass a small refactor.

### Pitch

Create a trivial loop for manual optimization. Trivial because it would be a loop running only one iteration as there is no iteration over optimizers. The `TrainingBatchLoop` would dynamically call either into `OptimizerLoop` or `ManualOptimization` depending on the user request. 

A POC is already here: #9266 

### Additional context

Please approve this RFC.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7938 
Follow up to #9191 
Fixes #9283 

Now that automatic optimization is separated into its own loop, we can finally drastically simplify the calls to manual optimization and move it to a separat loop class. 

TODO: 
- [x] make mypy happy

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -46,6 +40,7 @@ def __init__(self) -> None:
         # the current split index when the batch gets split into chunks in truncated backprop through time
         self.split_idx: Optional[int] = None
         self.optimizer_loop = OptimizerLoop()
+        self.manual_loop = ManualOptimization()
 
         self._warning_cache: WarningCache = WarningCache()
         self._hiddens: Optional[Tensor] = None
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:46]: I think now we can remove `_hiddens` in this class and instead track it inside the `OptimizerLoop` (already done) and `ManualOptimization`

Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:46]: ah, seems like it!

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:46]: actually, could you double check again what you had in mind? the life cycle of hidden state here is actually tied to the training batch loop if I'm not mistaken. 

I don't see it at the moment

Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py:46]: You are right.

File: pytorch_lightning/loops/batch/manual.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        return False
+
+    def reset(self) -> None:
+        raise NotImplementedError
+
+    def advance(self, *args: Any, **kwargs: Any) -> None:
+        """Unimplemented for manual optimization as there is no effective progression of the loop."""
+        raise NotImplementedError
+
+    def run(
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/manual.py:47]: Maybe we shouldn't use `run` for this and use `advance` where we set a flag like `self._done = True`.

Otherwise nobody would extend this loop.

And we might need `on_skip` and `reset` in the future

Comment (tchaton) [reply pytorch_lightning/loops/batch/manual.py:47]: Good point !

File: pytorch_lightning/loops/batch/manual.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        self._output: Optional[ResultCollection] = None
+
+    @property
+    def done(self) -> bool:
+        return self._done
+
+    def reset(self) -> None:
+        self._done = False
+
+    def advance(self, batch: Any, batch_idx: int, hiddens: Optional[Any] = None) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/manual.py:49]: kindly asking advice from typing experts, e.g. @carmocca or @justusschock: I'm getting the following: 

```bash
mypy pytorch_lightning/loops/batch/manual.py | grep manual 

pytorch_lightning/loops/batch/manual.py:49: error: Signature of "advance" incompatible with supertype "Loop"
```
Apparently 

`def advance(self, batch: Any, batch_idx: int, hiddens: Optional[Any] = None) -> None:`
is not a subtype of

`def advance(self, *args: Any, **kwargs: Any) -> None:`

What can we do about this?

File: pytorch_lightning/loops/batch/manual.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/manual.py]: I adjusted the loop structure to this. Note we also need to store the output and the hiddens

File: pytorch_lightning/loops/batch/manual.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        Args:
+            batch: the current tbptt split of the current batch
+            batch_idx: the index of the current batch
+            hiddens: the model's hidden state of the previous iteration
+
+        Returns:
+            post-processed outputs from the training step, or ``None`` if training step returned nothing
+        """
+        assert self.trainer is not None
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/manual.py:60]: reviewer: this nonsense is necessary to make mypy shut up complaining that trainer is None

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 5]
```diff
@@ -63,8 +58,9 @@ def optimizer_freq_cumsum(self) -> int:
             self._optimizer_freq_cumsum = np.cumsum(self.trainer.optimizer_frequencies)
         return self._optimizer_freq_cumsum
 
-    def connect(self, optimizer_loop: "Loop") -> None:
+    def connect(self, optimizer_loop: "Loop", manual_loop: ManualOptimization) -> None:
         self.optimizer_loop = optimizer_loop
+        self.manual_loop = ManualOptimization()
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:63]:
```suggestion
        self.manual_loop = manual_loop
```

File: pytorch_lightning/loops/batch/manual.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/manual.py:49]: https://stackoverflow.com/a/51003172

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/batch/manual.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/manual.py]: I disagree (maybe because I don't understand fully) but the design is intentionally this way. The flexibility must be there so a user can implement their loop accepting whatever args they want in run(). In this sense the stackoverflow answer does not apply to our needs.

Comment (carmocca) [reply pytorch_lightning/loops/batch/manual.py]: mypy fails because it violates the Liskov Substitution Principle.

Meaning that if an object of a superclass loop is replaced with an object of a subclass loop, it could break because the subclass doesn't take any number of arguments as the superclass signature would indicate.

File: pytorch_lightning/loops/batch/manual.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        self._output: Optional[ResultCollection] = None
+
+    @property
+    def done(self) -> bool:
+        return self._done
+
+    def reset(self) -> None:
+        self._done = False
+
+    def advance(self, batch: Any, batch_idx: int, hiddens: Optional[Any] = None) -> None:  # type: ignore [override]
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/manual.py:49]:
```suggestion
    def advance(self, batch: Any, batch_idx: int, hiddens: Optional[Any] = None) -> None:  # type: ignore[override]
```
@carmocca IDE shows me an error with the space

Comment (carmocca) [reply pytorch_lightning/loops/batch/manual.py]: You are right, i've been writing it wrong the whole time then 😂

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 7]
```diff
@@ -45,6 +39,7 @@ def __init__(self) -> None:
         # the current split index when the batch gets split into chunks in truncated backprop through time
         self.split_idx: Optional[int] = None
         self.optimizer_loop = OptimizerLoop()
+        self.manual_loop = ManualOptimization()
```
Comment (Borda) [inline pytorch_lightning/loops/batch/training_batch_loop.py:42]: this is a bit confusing, this looks like we have two loops for the optimizer? how they interact?

Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:42]: they don't interact. automatic optimization has an optimizer loop and manual optimization does not. so either one or the other is active, never both

Comment (Borda) [reply pytorch_lightning/loops/batch/training_batch_loop.py:42]: so would it be cleaner to have also assigned only one of them?

Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:42]: @Borda Do you mean something like this: 

```py
optimization_loop = OptimizerLoop() if automatic_optimization else ManualOptimization()
```

Comment (Borda) [reply pytorch_lightning/loops/batch/training_batch_loop.py:42]: yes :rabbit:

Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:42]: this information is not available at the time of instantiation. manual_optimization True/False is defined on the LightningModule, therefore this will only be known at the time when fit is called.

Same logic applies also to TBPTT where we have the batch loop even when we don't actually know that `model.truncated_bptt_steps > 0`.

Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:42]: IMO, I prefer the logic introduced by @awaelchli, as the code is well separated.
It would be possible in the future to get more lower-level loop.

File: pytorch_lightning/loops/batch/manual.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    def advance(self, batch: Any, batch_idx: int, hiddens: Optional[Any] = None) -> None:  # type: ignore[override]
+        """Performs the training step for manual optimization.
+
+        Args:
+            batch: the current tbptt split of the current batch
+            batch_idx: the index of the current batch
+            hiddens: the model's hidden state of the previous iteration
+
+        Returns:
+            post-processed outputs from the training step, or ``None`` if training step returned nothing
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/manual.py:57]: This method does not return

File: pytorch_lightning/loops/batch/manual.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        self._done = True
+        self._hiddens = hiddens
+        self._output = result_collection
+
+    def on_run_end(self) -> Optional[Tuple[ResultCollection, Optional[Any]]]:
+        hiddens = self._hiddens
+        output = self._output
+        self._hiddens, self._output = None, None  # free memory
+        if output is None:
+            return None
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/manual.py:91]: Why not return the tuple directly?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 10]
```diff
@@ -132,10 +132,11 @@ def advance(self, batch, batch_idx):
...
                 self.batch_outputs[k].extend(batch_outputs[k])
         else:
-            # in manual optimization, there is no looping over optimizers
-            result = self._run_optimization(batch_idx, split_batch)
-            if result:
-                self.batch_outputs[0].append(deepcopy(result.result_collection))
+            # in manual optimization, hand over execution to the ManualOptimization loop
+            output = self.manual_loop.run(split_batch, batch_idx, self._hiddens)
+            if output:
+                result, self._hiddens = output
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:138]: Who would use `manual_optimization` and tbptt?

I don't think we have a single test with manual optimization and tbptt enabled. Can you add one?

File: pytorch_lightning/loops/batch/manual.py (hunk 9 above)
Comment (justusschock) [reply pytorch_lightning/loops/batch/manual.py:91]: I think it's easier to check if the output of a function is None than to check it for each item in the tuple :)

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/batch/manual.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/manual.py:91]: Should manual optimization allow returning `{loss: None, hiddens: something()}`?

Should anything returned by manual optimization be considered and `extra` and treated as such? What about the hiddens in that case? And the loss?

I think we should sit down and properly define these things because the integration of all these components is currently undefined and untested.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: I updated one test case to use manual optimization and automatic optimization.

File: pytorch_lightning/loops/batch/manual.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/loops/batch/manual.py]: moved the doc to the on_run_end where we return.

File: tests/models/test_truncated_bptt.py [hunk 11]
```diff
@@ -14,82 +14,97 @@
...
+        self.sequence_size = sequence_size
+        self.batch_size = batch_size
+        self.automatic_optimization = True
+
+        self.example_input_array = torch.randn(5, truncated_bptt_steps)
+        self.layer = torch.nn.Linear(in_features=truncated_bptt_steps, out_features=truncated_bptt_steps)
+        self.test_hidden = None
+
+    def training_step(self, batch, batch_idx, hiddens):
+        assert hiddens == self.test_hidden, "Hidden state not persistent between tbptt steps"
```
Comment (carmocca) [inline tests/models/test_truncated_bptt.py:39]: Can you add this hiddens assertion to the manual optimization model?

Comment (awaelchli) [reply tests/models/test_truncated_bptt.py:39]: it's called by the super().training_step already :)

Comment (carmocca) [reply tests/models/test_truncated_bptt.py:39]: Oops, missed it!

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 12]
```diff
@@ -145,89 +145,6 @@ def num_active_optimizers(self, batch_idx: Optional[int] = None) -> int:
...
-        """
-        # TODO: replace call through closure by direct call (manual optimization)
-        closure = self._make_closure(split_batch, batch_idx, self._hiddens)
-        closure()
-        result = closure.get_result()
-
-        if result:
-            # if no result, user decided to skip optimization
-            # otherwise update running loss + reset accumulated loss
-            self._update_running_loss(result.loss)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:167]: technically this should not have been removed, ... sigh
but yeah, we can see what happens with #9372
